clang/lib/AST/MicrosoftMangle.cpp

```diff
@@ -2902,6 +2902,17 @@ void MicrosoftCXXNameMangler::mangleCallingConvention(CallingConv CC) {
       else
         Out << "w";
       break;
+    case CC_OpenCLKernel:
```

Comment:
This looks ok to me but @premanandrao can you also please take a look?

Reply:
I have a lot to learn in this area. I don't understand why OpenCL is relevant for the added test case. Is OpenCL always used for `-fsycl-is-native-cpu`?

Mangling also seems to be missing for IA64 ABIs. See `CXXNameMangler::getCallingConvQualifierName()` in `clang/lib/AST/ItaniumMangle.cpp`. Why isn't a similar change needed there? The test exercises Linux as well.

Reply:
Has anyone tried to determine what, if any, mangling is used by MSVC for OpenCL kernels?

Reply:
>>I have a lot to learn in this area. I don't understand why OpenCL is relevant for the added test case. Is OpenCL always used for -fsycl-is-native-cpu?

My understanding was that this is this calling convention for generated kernel. I think that when using `-fsycl-is-native-cpu` device compilation is done with the host triple and this results in us entering MS mangler for the kernel. @uwedolinsky did I get that correct?

>Mangling also seems to be missing for IA64 ABIs. See CXXNameMangler::getCallingConvQualifierName() in clang/lib/AST/ItaniumMangle.cpp. Why isn't a similar change needed there? The test exercises Linux as well.

This is a good point. Why is this not an issue for Linux?

Reply:
> I don't understand why OpenCL is relevant for the added test case. 

OpenCL is not relevant for this test,  but since the frontend attaches the clang `CC_OpenCLKernel` convention  (independent from C++ ABI and target) to the kernel declaration [here](https://github.com/intel/llvm/blob/8b52f12f04c228eebf6d5a1c7f7ba9f74d0ec5f0/clang/lib/Sema/SemaSYCL.cpp#L2321), we need to handle that convention in the currently used name mangler (which is the MS mangler when targeting Windows).

> Is OpenCL always used for `-fsycl-is-native-cpu`?

`-fsycl-is-native-cpu` does not use or enable OpenCL in compiler or runtime. `NativeCPU` has no dependency on OpenCL. 
 
> Mangling also seems to be missing for IA64 ABIs. See `CXXNameMangler::getCallingConvQualifierName()` in `clang/lib/AST/ItaniumMangle.cpp`. Why isn't a similar change needed there? 

`CXXNameMangler::getCallingConvQualifierName()` already handles the `CC_OpenCLKernel` and other conventions [here](https://github.com/intel/llvm/blob/8b52f12f04c228eebf6d5a1c7f7ba9f74d0ec5f0/clang/lib/AST/ItaniumMangle.cpp#L3281) (without providing mangling), but this doesn't even seem to be called on SYCL kernels when targeting the IA64 ABI, so that problem will not occur on that ABI.     

> The test exercises Linux as well.

We wanted the test to also cover other ABIs and different combinations of cpu triples (for possible cross compilation) to ensure they don't have similar problems. Some of these invocations may not exercise the fix but are still worth testing.

Reply:
> My understanding was that this is the calling convention for kernel. I think that when using `-fsycl-is-native-cpu` device compilation is done with the host triple and this results in us entering MS mangler for the kernel. @uwedolinsky did I get that correct?

Yes, that's correct. Whenever `-fsycl-is-native-cpu` targets device code to one of the cpu triples for Windows, the MSABI/Mangler will be used. This is also important for being able to debug device code in Visual Studio.
 
> > Mangling also seems to be missing for IA64 ABIs. See CXXNameMangler::getCallingConvQualifierName() in clang/lib/AST/ItaniumMangle.cpp. Why isn't a similar change needed there? The test exercises Linux as well.
> 
> This is a good point. Why is this not an issue for Linux?

Please see my comment above, but essentially this convention is already handled by the Itanium ABI implementation which doesn't have that issue.

Reply:
> Has anyone tried to determine what, if any, mangling is used by MSVC for OpenCL kernels?

MSVC doesn't seem to support the OpenCL qualifiers like `__kernel`, `__global` etc. So it can't really compile OpenCL kernels (and mangle their names) unless these OpenCL qualifiers are #defined away or to something MSVC understands so it can apply its usual mangling.

Reply:
Thank you for explaining. I think I'm convinced this change is ok.

Reply:
It still isn't clear to me that this is the right change. Why is the `CC_OpenCLKernel` attached to SYCL kernels in the first place? Can we just stop attaching them to SYCL kernels? I think we should understand why they are needed before choosing to ignore them.

MSVC doesn't directly support the OpenCL language. From what I can tell, OpenCL kernels are always resolved by a call to `clCreateKernel()` that passes the name of the kernel as a string. Overloaded kernels are supported (at least by some implementations) by allowing a custom mangled name to be specified via an attribute. It seems any kind of mangling is therefore implementation dependent and thus, the Microsoft ABI doesn't specify one.

Reply:
> It still isn't clear to me that this is the right change. Why is the `CC_OpenCLKernel` attached to SYCL kernels in the first place? 

Probably because of historical reasons since SYCL initially targeted OpenCL, many data structures and identifiers may have been inspired by it and SYCL kernels were implemented as OpenCL kernels. Also, querying the calling convention is one way to identify kernels.  

> Can we just stop attaching them to SYCL kernels? 

Probably, but that would likely lead to a much larger, potentially intrusive SYCL frontend change I mentioned previously and I was trying to avoid that. If NativeCPU is the only SYCL target that has this issue, and there are no other issues elsewhere with using this convention "under the hood", it may not be worth making that large a change and instead it may be easier to just handle the assert. 

> MSVC doesn't directly support the OpenCL language. From what I can tell, OpenCL kernels are always resolved by a call to `clCreateKernel()` that passes the name of the kernel as a string. Overloaded kernels are supported (at least by some implementations) by allowing a custom mangled name to be specified via an attribute. It seems any kind of mangling is therefore implementation dependent and thus, the Microsoft ABI doesn't specify one.

Please note that this PR does *not* specify any MS mangling, it merely prevents the assert in the MS mangler triggered for the `CC_OpenCLKernel` convention.

Reply:
Thank you for being responsive to my comments. I won't hold this review up further; the change seems unlikely to cause any problems in the short term. However, we will have to revisit all of this as part of the SYCL upstreaming effort as the reuse of `CC_OpenCLKernel` for SYCL here is quite confusing.

Reply:
Yes, if the clang infrastructure for OpenCL isn't used to implement SYCL kernels (not even when targeting OpenCL), then this "reuse" of `CC_OpenCLKernel` could probably be avoided. It may be possible to instead use the information from  `SYCLKernelAttr` or other SYCL-related clang attributes that are currently attached to SYCL kernels in the AST. But such potentially larger refactoring is probably better done as a separate PR.